refactor(replace nats): add directory structure and proto for pool

Abhinandan-Purkait opened the pull request:
Add pool proto for the initial phase of implementation of grpc to replace nats.

paulyoong commented on control-plane/grpc/protos/pool.proto:
```diff
@@ -0,0 +1,133 @@
+  Metadata metadata = 2;
+}
+
+message Metadata {
```
I think we should include a comment for this message.

Abhinandan-Purkait replied:
Added comment.

paulyoong commented on control-plane/grpc/protos/pool.proto:
```diff
@@ -0,0 +1,133 @@
+import "google/protobuf/empty.proto";
+import "google/protobuf/wrappers.proto";
+
+package grpc;
```
Do we want the package to be generically named? Will other proto file (i.e. volume.proto) also be package grpc? Alternatively we could call it pool-grpc. What do you think?

Abhinandan-Purkait replied:
Have split into directory versioned directories and changed it to `pool` package.

tiagolobocastro commented on control-plane/grpc/protos/pool.proto:
```diff
@@ -0,0 +1,133 @@
+}
+
+// PoolLabel is the type for the labels
+message PoolLabel {
```
Should this be a common type, eg: StringMapValue?

Abhinandan-Purkait replied:
Added the type.

tiagolobocastro commented on control-plane/grpc/protos/pool.proto:
```diff
@@ -0,0 +1,133 @@
+
+// PoolLabel is the type for the labels
+message PoolLabel {
+  map<string, string> labels = 1;
```
and then this could be called value for example

tiagolobocastro commented on control-plane/grpc/protos/pool.proto:
```diff
@@ -0,0 +1,133 @@
+}
+
+// Service for managing storage pools
+service CoreGrpc {
```
Should this be named PoolGrpc?

tiagolobocastro commented on control-plane/grpc/protos/pool.proto:
```diff
@@ -0,0 +1,133 @@
+
+// Service for managing storage pools
+service CoreGrpc {
+  rpc CreatePool (CreatePoolRequest) returns (Pool) {}
```
Do we need to define an error type? On the bus we currently return `ReplyError` in case of an error.
Then that be sent on the .details of the grpc `Status`?

Abhinandan-Purkait replied:
the code generator translates the return type to `Result<Pool, tonic::Status>`. 
`tonic::Status` -> ref: https://github.com/hyperium/tonic/blob/master/tonic/src/status.rs
After a discussion with @paulyoong,  we have added a composite type for ex :-
```
message CreatePoolReply {
  oneof reply {
    Pool pool = 1;
    ReplyError error = 2;
  }
}
```
with this, we can propagate internal error other than grpc error(denoted by `tonic::status`)
And the current code generated would look somewhat like this:- `Result<CreatePoolReply, tonic::Status>`

tiagolobocastro commented on control-plane/grpc/protos/common.proto:
```diff
@@ -0,0 +1,11 @@
+
+package grpc;
+
+// Enum defining the various states that a resource spec can be in.
```
s/states/statuses ?

paulyoong commented on control-plane/grpc/proto/pool/v1/pool.proto:
```diff
@@ -0,0 +1,157 @@
+  string pool_id = 2;
+}
+
+// Get all the pools from based on the args
```
Nit: Maybe this should say something like "Get all pools based on the filter criteria"

tiagolobocastro commented on control-plane/grpc/src/lib.rs:
```diff
@@ -0,0 +1,7 @@
+pub(crate) mod common {
```
Should these be `pub`?
Should we put some doc comments above these?

Abhinandan-Purkait commented:
FYI: I have changed the structure a bit, as there was issue keeping `resource/version` structure. So currently I have moved everything under v1, and that way it seems to work, also I have added the errors. Needs one more round of review.

paulyoong commented on control-plane/grpc/proto/v1/misc/common.proto:
```diff
@@ -0,0 +1,86 @@
+  AlreadyShared = 20;
+  NotPublished = 21;
+  AlreadyPublished = 22;
+  Deleting_ = 23;
```
If possible can we get rid of the underscore suffix. Maybe we could rename it to `IsDeleting`?

paulyoong commented on control-plane/grpc/proto/v1/pool/pool.proto:
```diff
@@ -0,0 +1,151 @@
+  Metadata metadata = 2;
+}
+
+// uuid and the status of pool for the control plane use
```
I would suggest rewording `status of pool` to `status of the pool spec` otherwise it may be read as the health of the pool (i.e. online, degraded etc.) rather than whether the pool is being created/deleted. It's the distinction between `SpecStatus` and `PoolStatus`.

paulyoong commented on control-plane/grpc/proto/v1/pool/pool.proto:
```diff
@@ -0,0 +1,151 @@
+  Online = 1;
+  // the pool has experienced a failure but can still function
+  Degraded = 2;
+  // the pool is completely inaccessible
```
Nit: I wonder if we should say that the pool is unable to be used rather than inaccessible. It may be accessible, i.e. we can connect to it, but the disks may be faulty.

Abhinandan-Purkait commented:
bors merge
